### DISCUSSION TITLE
Possible to group the drawings | Creating a complex drawing in PyMuPDF

### DISCUSSION THREAD

**Discussion by binodil** [pymupdf/PyMuPDF#2507] [Looking for help]
Hello there,

**I got all objects in a specific area in pdf V1. (Like getting the drawings in the center location of the pdf file). 
I wonder if there is a way to connect all simple drawings like rectangles, circles, lines as complex object.** 

## Why I need this complex object?
I want to quickly relocate this complex object as one object, instead of moving every simple drawing one by one.  At the same time, I want to maintain the drawings metadata. Initially, I  was planning to use `pixmap`, but as far as I know it eliminate the drawings properties.

**Comment by JorjMcKie** [comment]
The vector graphics extraction (fairly) faithfully reports the graphics commands in the page's `/Contents`. Whether they belong somehow together, beyond the single drawing paths is impossible to detect.
You can of course try to group yourself by looking at which path rectangles overlap each other, etc. I have script that does this somewhere ...
So you may end up with a rectangle that envelops a list single paths. Then take this list for further processing.

**Reply by JorjMcKie** [reply]
[Here](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-vector-graphics/separate-figures.py) is that mentioned script.

**Reply by binodil** [reply]
_Thanks for your information. 
I have edit the code above. It works and does main job._ 

I wonder if you can help me with following issue.
Assume, I have a fitz.Rect(0,0,10,12) which is threshold which filters the object based on their location to this **rect**.
In some cases, some drawings are located in the middle of the threshold => so their are not included or half included. 

What advice would you suggest to the overcome this problem. I hope I could explain the code. 
I think that instead of following code, I must modify the `r` variable. 
`
for p in page.get_drawings():
            w = p["width"]  # thickness of the border line
            r = p["rect"] + (-w, -w, w, w)  # enlarge each rectangle by width value
            for i in range(len(new_rects)):
`
 
Looking forward to hearing you!

**Reply by JorjMcKie** [reply]
Hard to say without seeing your page.
But you of course can join rectangles and thus create a larger one that includes bot `r = r1 | r2`, with the rectangle "union" operator "|". So you can start with a "threshold" rectangle and join every path that intersects it like that:
```python
for path in paths:
    prect = path["rect"]
    if prect.insersects(threshold):
        threshold |= prect
```
Instead of simply checking "intersects yes / no", you can also compute whether a large part of prect is contained: `if abs(prect & threshold) >= abs(prect) * 0.7: ...`. This checks whether 70% of the prect area is contained in threshold.

If that makes sense in your case is up to you to decide.
